[Composable] Use non-reentrant generator, remove reentrant

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106003
* #106002
* #106001
* #106000
* #105999
* #105995
* #105991
* #105207
* #105206
* #105205
* #105180
* __->__ #105176

Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

rohan-varma committed:
[Composable] Use non-reentrant generator, remove reentrant

Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Use non-reentrant generator, remove reentrant"


Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Use non-reentrant generator, remove reentrant"


Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/checkpoint_activation.py:
```diff
@@ -214,63 +62,39 @@ def checkpoint(module: nn.Module, *, use_reentrant: bool = True) -> nn.Module:
     module.register_forward_pre_hook(forward_pre_hook)
-    # Use prepend to make sure we restore the original grad enabled state right
-    # after the module forward invocation.
-    module.register_forward_hook(forward_hook, prepend=True)
+    module.register_forward_hook(forward_hook, prepend=True, always_call=True)
```
Given that we are not passing kwargs into the hooks, I think we will still need to fix for supporting kwargs.

rohan-varma replied:
sounds good, we can fix this in the follow up diffs?

awgu commented on test/distributed/_composable/test_checkpoint.py:
```diff
@@ -11,8 +11,6 @@
 from torch.distributed._composable import checkpoint
```
Remove the `use_reentrant` kwarg in `test_checkpoint_fsdp_submodules_non_reentrant()`:
```
test_model.u1.seq = checkpoint(test_model.u1.seq, use_reentrant=use_reentrant)
```

awgu commented on test/distributed/_composable/test_checkpoint.py:
```diff
@@ -184,17 +174,59 @@ def test_multi_args(self, use_reentrant: bool):
+        # Should raise in forward recomputation
+        out = m(inp).sum()
+
+        with self.assertRaises(RuntimeError):
```
```suggestion
        with self.assertRaisesRegex(RuntimeError, "foo"):
```
(or factor the runtime error message into a local variable and use that)

awgu commented on test/distributed/_composable/test_checkpoint.py:
```diff
@@ -184,17 +174,59 @@ def test_multi_args(self, use_reentrant: bool):
 
+    def test_raise_in_forward(self):
+
+        class MyModel(torch.nn.Module):
```
nit: For AC specifically, it can be good to think about applying it to a 'module' as opposed to a 'model' because realistically we would never checkpoint the entire model.

rohan-varma replied:
sounds good, will refactor this into a container

awgu commented on test/distributed/_composable/test_checkpoint.py:
```diff
@@ -184,17 +174,59 @@ def test_multi_args(self, use_reentrant: bool):
         for p1, p2 in zip(net1.parameters(), net2.parameters()):
             self.assertEqual(p1.grad, p2.grad)
 
+    def test_raise_in_forward(self):
```
nit: Maybe we could call this test `test_clears_state_on_error_in_forward()` (or similar) or somehow otherwise clarify that this is testing that `checkpoint`-specific state is cleared on an error in forward?

awgu commented on torch/distributed/_composable/checkpoint_activation.py:
```diff
@@ -214,63 +62,39 @@ def checkpoint(module: nn.Module, *, use_reentrant: bool = True) -> nn.Module:
-                checkpoint.state(module).saved_tensor_hooks.__exit__()
-                checkpoint.state(module).saved_tensor_hooks = None
+                raise RuntimeError(
+                    "Expected non-reentrant activation checkpoint generator to be exhausted, but it was not!"
```
What happens if we hit this error? What do we do?

rohan-varma replied:
I'm not sure if this can happen other than an implementation change / bug where the AC generator can be `next`'ed more than twice. I guess we should tell the user to file a repro?

awgu replied:
Sounds good!

awgu reviewed (approved):
Stamp to unblock -- please fix the failing unit test and lint before landing :)

rohan-varma committed:
Update on "[Composable] Use non-reentrant generator, remove reentrant"


Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Use non-reentrant generator, remove reentrant"


Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Composable] Use non-reentrant generator, remove reentrant"


Removes reentrant support for the composable checkpoint, as
non-reentrant is the recommended approach and we should use this when rolling
out composable checkpoint API.

Also removes the standalone implementation for non-reentrant and instead uses
the generator from below diff to reuse the original implemenetation.

Differential Revision: [D47451375](https://our.internmc.facebook.com/intern/diff/D47451375/)

[ghstack-poisoned]
